### PR TITLE
Basic api to execute casadm commands

### PR DESCRIPTION
**Pull request by KamilLepek**
Added command line help test
Created base test prepare which creates test properties
Check if cas is installed in basic test prepare and install if needed

Signed-off-by: Kamil Lepek <kamil.lepek94@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/open-cas/test-framework/13)

### CODE REVIEW

File: api/casadm.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+#
+# Copyright(c) 2019 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+#
+
+from .cli import *
+from test_package.test_properties import TestProperties
+
+
+def help(shortcut: bool = False):
```
Comment (Deixx) [inline api/casadm.py]: I think some prefix for CAS commands would be useful, calling just 'help()' or 'start()' is not telling much

File: api/casadm.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    return TestProperties.executor.execute(help_cmd(shortcut))
+
+
+# TODO:In the future cache_dev will propably be more complex than string value
```
Comment (Deixx) [inline api/casadm.py]: probably

File: api/casadm.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+#
```
Comment (Deixx) [inline api/casadm.py]: Is the folder generally for APIs or just CAS API? If the latter then the name could reflect that. cas_api or maybe apis/cas if more APIs are to be placed here.

File: api/cli.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+LOGGER = logging.getLogger(__name__)
+
+casadm_bin = "casadm"
+
+
+def add_core_cmd(cache_id: int, core_dev: str, core_id: int = None, shortcut: bool = False):
+    command = f" -A -i {str(cache_id)} -d {core_dev}" if shortcut \
+        else f" --add-core --cache-id {str(cache_id)} --core-device {core_dev}"
+    if core_id is not None:
+        command += + " -j " + str(core_id) if shortcut else " --core-id " + str(core_id)
```
Comment (Deixx) [inline api/cli.py]: I don't think '+= +' is intended.

File: api/cli.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+def remove_core_cmd(cache_id: int, core_id: int, force: bool = False, shortcut: bool = False):
+    command = f" -R -i {cache_id} -j {core_id}" if shortcut \
+        else f" --remove-core --cache-id {cache_id} --core-id {core_id}"
+    if force:
+        command += " -f" if shortcut else " --force"
+    return casadm_bin + command
+
+
+def help_cmd(shortcut: bool = False):
+    return casadm_bin + " -H" if shortcut else casadm_bin + " --help"
```
Comment (Deixx) [inline api/cli.py]: repeating variable catches my eye - I'd prefer:
return casadm_bin + (" -H" if shortcut else " --help")

File: api/cli.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+
+def start_cmd(cache_dev: str, cache_mode=None, cache_line_size=None,
+              cache_id=None, force=False, load=False, shortcut=False):
+    command = " -S" if shortcut else " --start-cache"
+    command += " -d " + cache_dev if shortcut else " --cache-device " + cache_dev
```
Comment (Deixx) [inline api/cli.py]: same here and in many places below ;)

File: cas_configuration/cache_config.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    LINE_4KiB = 4096
+    LINE_8KiB = 8192
+    LINE_16KiB = 16384
+    LINE_32KiB = 32768
+    LINE_64KiB = 65536
+    DEFAULT = LINE_4KiB
+
+
+# TODO: Use case for this will be to iterate over configurations (kernel params such as
+# TODO: such as io schedler, metadata layout) and prepare env before starting cache
```
Comment (Deixx) [inline cas_configuration/cache_config.py]: scheduler

File: test_package/cli/test_cli_help.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
+#
+# Copyright(c) 2019 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+#
+
+
+import logging
+import pytest
+from api import casadm
+from test_package.conftest import base_prepare as base_prepare
```
Comment (Deixx) [inline test_package/cli/test_cli_help.py]: base_prepare as base_prepare?

File: test_package/cli/test_cli_start_stop.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
+#
+# Copyright(c) 2019 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+#
+
+
+import logging
+import pytest
+from api import casadm
+from test_package.conftest import base_prepare as base_prepare
```
Comment (Deixx) [inline test_package/cli/test_cli_start_stop.py]: base_prepare as base_prepare?

File: cas_configuration/cache_config.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+# Copyright(c) 2019 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+#
+
+from enum import IntEnum
+
+
+# TODO: change to some sort of Size class when it is implemented
+class CacheLineSize(IntEnum):
+    LINE_4KiB = 4096
```
Comment (Deixx) [inline cas_configuration/cache_config.py]: Size is implemented :)

File: test_package/conftest.py [hunk 11]
```diff
@@ -15,15 +16,21 @@
...
 
 @pytest.fixture()
 def prepare_and_cleanup(request):
     """
-    This fixture returns the dictionary, which contains DUT ip, IPMI, spider, list of disks
+    This fixture returns the dictionary, which contains DUT ip, IPMI, spider, list of disks.
+    This fixture also returns the executor of commands
     """
 
     # There should be dut config file added to config package and
```
Comment (Deixx) [inline test_package/conftest.py]: change --config to --dut-config in the next line

File: test_package/conftest.py [hunk 12]
```diff
@@ -64,5 +71,34 @@ def prepare_and_cleanup(request):
...
 def pytest_addoption(parser):
-    parser.addoption("--config", action="store", default="config/configuration.py")
+    parser.addoption("--dut-config", action="store", default="None")
+    parser.addoption("--remote", action="store", default="origin")
     parser.addoption("--repo-tag", action="store", default="master")
+    parser.addoption("--force-reinstall", action="store_true", default="False")
+    # TODO: investigate whether it is possible to pass the last param as bool
+
+
+def get_remote():
```
Comment (Deixx) [inline test_package/conftest.py]: I'm not sure if non fixture related methods should be placed here, but I won't fight for it ;)

File: test_package/example_test.py [hunk 13]
```diff
@@ -5,10 +5,8 @@
 
 import logging
 import pytest
-import time
-from utils.dut import Dut
 from test_package.test_properties import TestProperties
-
+from test_package.conftest import base_prepare as base_prepare
```
Comment (Deixx) [inline test_package/example_test.py]: great alias! :)

File: api/casadm.py (hunk 1 above)
Comment (KamilLepek) [reply api/casadm.py]: The tests I wrote import casadm as package so the usage of these methods is casadm.help() or casadm.start() which is telling enough in my opinion. If we use it like that in the tests then it shouldn't be a problem. However if you have some other idea just let me know :)

File: api/casadm.py (hunk 2 above)
Comment (KamilLepek) [reply api/casadm.py]: Done

File: api/casadm.py (hunk 3 above)
Comment (KamilLepek) [reply api/casadm.py]: Done

File: api/cli.py (hunk 4 above)
Comment (KamilLepek) [reply api/cli.py]: Done

File: api/cli.py (hunk 5 above)
Comment (KamilLepek) [reply api/cli.py]: Done

File: api/cli.py (hunk 6 above)
Comment (KamilLepek) [reply api/cli.py]: Done

File: cas_configuration/cache_config.py (hunk 7 above)
Comment (KamilLepek) [reply cas_configuration/cache_config.py]: Done

File: cas_configuration/cache_config.py (hunk 10 above)
Comment (KamilLepek) [reply cas_configuration/cache_config.py]: Done

File: test_package/cli/test_cli_help.py (hunk 8 above)
Comment (KamilLepek) [reply test_package/cli/test_cli_help.py]: refactor relic :), done

File: test_package/cli/test_cli_start_stop.py (hunk 9 above)
Comment (KamilLepek) [reply test_package/cli/test_cli_start_stop.py]: Done

File: test_package/conftest.py (hunk 11 above)
Comment (KamilLepek) [reply test_package/conftest.py]: Done

File: test_package/example_test.py (hunk 13 above)
Comment (KamilLepek) [reply test_package/example_test.py]: Done

File: test_package/conftest.py (hunk 12 above)
Comment (KamilLepek) [reply test_package/conftest.py]: I agree, however for now I'll just add a _TODO_

File: api/casadm.py (hunk 1 above)
Comment (Deixx) [reply api/casadm.py]: If you could enforce it to be used scritclty like in your tests then it'd be fine. But 'from api.cas.casadm import help' will allow for usage of methods without mentioning casadm, and then calling help() can be quite confusing.

Comment (KamilLepek) [reply api/casadm.py]: Done

File: api/cas/casctl.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+def help(shortcut: bool = False):
+    return TestProperties.executor.execute(ctl_help(shortcut))
+
+
+def start():
+    return TestProperties.executor.execute(ctl_start())
+
+
+def stop():
```
Comment (Deixx) [inline api/cas/casctl.py]: has parameter flush

File: api/cas/casctl.py [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    return TestProperties.executor.execute(ctl_stop())
+
+
+def init():
```
Comment (Deixx) [inline api/cas/casctl.py]: has parameter force

File: api/cas/casctl.py (hunk 14 above)
Comment (KamilLepek) [reply api/cas/casctl.py]: Done

File: api/cas/casctl.py (hunk 15 above)
Comment (KamilLepek) [reply api/cas/casctl.py]: Done

File: api/cas/casadm.py [hunk 16]
```diff
@@ -86,3 +152,48 @@ def parse_list_caches():
...
+                     io_class_id: int = None, filter: StatsFilter = None,
+                     output_format: OutputFormat = None, shortcut: bool = False):
+    _core_id = None if core_id is None else str(core_id)
+    _io_class_id = None if io_class_id is None else str(io_class_id)
+    output = TestProperties.executor.execute(
+        print_statistics_cmd(
+            str(cache_id), _core_id, per_io_class, _io_class_id, filter, output_format, shortcut))
+    if output.exit_code != 0:
+        raise Exception(
+            f"Format command failed. stdout: {output.stdout} \n stderr :{output.stderr}")
```
Comment (Deixx) [inline api/cas/casadm.py]: Wrong command description

File: api/cas/casadm.py [hunk 17]
```diff
@@ -86,3 +152,48 @@ def parse_list_caches():
...
+        raise Exception(
+            f"Set cache mode command failed. stdout: {output.stdout} \n stderr :{output.stderr}")
+    return output
+
+
+def load_io_classes(cache_id: int, file: str, shortcut: bool = False):
+    output = TestProperties.executor.execute(load_io_classes_cmd(str(cache_id), file, shortcut))
+    if output.exit_code != 0:
+        raise Exception(
+            f"Load ioclass command failed. stdout: {output.stdout} \n stderr :{output.stderr}")
```
Comment (Deixx) [inline api/cas/casadm.py]: I'd change description to 'Load IO class configuration...'

File: api/cas/casadm.py [hunk 18]
```diff
@@ -86,3 +152,48 @@ def parse_list_caches():
...
+    return output
+
+
+def list_io_classes(cache_id: int, output_format: OutputFormat, shortcut: bool = False):
+    output = TestProperties.executor.execute(
+        list_io_classes_cmd(str(cache_id), output_format, shortcut))
+    if output.exit_code != 0:
+        raise Exception(
+            f"List ioclass command failed. stdout: {output.stdout} \n stderr :{output.stderr}")
```
Comment (Deixx) [inline api/cas/casadm.py]: IO class

File: api/cas/casadm.py [hunk 19]
```diff
@@ -86,3 +152,48 @@ def parse_list_caches():
             parsed_output["cores"] = \
                 {args[1]: {"path": args[2], "state": args[3], "device": args[5]}}
     return parsed_output
+
+
+def print_statistics(cache_id: int, core_id: int = None, per_io_class: bool = False,
+                     io_class_id: int = None, filter: StatsFilter = None,
```
Comment (Deixx) [inline api/cas/casadm.py]: filter can be joined, for example: -f conf,usage

File: api/cas/cli.py [hunk 20]
```diff
@@ -0,0 +1,255 @@
...
+    return casadm_bin + command
+
+
+def set_param_cutoff_cmd(cache_id: str, core_id: str = None, threshold: str = None,
+                         policy: str = None, shortcut: bool = False):
+    add_params = ""
+    if core_id is not None:
+        add_params += (" -j " if shortcut else " --core-id ") + core_id
+    if threshold is not None:
+        add_params += (" --t " if shortcut else " --threshold ") + threshold
```
Comment (Deixx) [inline api/cas/cli.py]: -t

File: api/cas/casadm.py [hunk 21]
```diff
@@ -0,0 +1,318 @@
...
+    if output.exit_code != 0:
+        raise Exception(
+            f"Failed to remove detached core. stdout: {output.stdout} \n stderr :{output.stderr}")
+    return output
+
+
+def reset_counters(cache_id: int, core_id: int = None, shortcut: bool = False):
+    _core_id = None if core_id is None else str(core_id)
+    output = TestProperties.executor.execute(
+        reset_counters(cache_id=str(cache_id), core_id=_core_id, shortcut=shortcut))
```
[the remark this answers is not in the document]
Comment (KamilLepek) [reply api/cas/casadm.py]: Done

File: api/cas/casadm.py [hunk 22]
```diff
@@ -0,0 +1,318 @@
...
+    casctl_stop()
+    output = list_caches()
+    if "No caches running" not in output.stdout:
+        raise Exception(
+            f"Error while stopping caches. stdout: {output.stdout} \n stderr :{output.stderr}")
+
+
+def parse_list_caches():
+    parsed_output = {"caches": {}, "cores": {}}
+    lines = list_caches("csv").stdout.split('\n')
```
[the remark this answers is not in the document]
Comment (KamilLepek) [reply api/cas/casadm.py]: Done

File: api/cas/cli.py (hunk 20 above)
Comment (KamilLepek) [reply api/cas/cli.py]: Done

File: api/cas/casadm.py [hunk 23]
```diff
@@ -0,0 +1,319 @@
...
+def parse_list_caches():
+    parsed_output = {"caches": {}, "cores": {}}
+    lines = list_caches(OutputFormat.csv).stdout.split('\n')
+    for line in lines:
+        args = line.split(',')
+        if args[0] == "cache":
+            parsed_output["caches"] =\
+                {args[1]: {"path": args[2], "state": args[3], "mode": args[4]}}
+        elif args[0] == "core":
+            parsed_output["cores"] = \
```
[the remark this answers is not in the document]
Comment (KamilLepek) [reply api/cas/casadm.py]: Fixed for cache and core. Key for core was not exclusive so I changed the key from core id to exposed device name (e.g cas1-1)

File: api/cas/casadm.py [hunk 24]
```diff
@@ -0,0 +1,319 @@
...
+            f"Getting acp cleaning policy params failed."
+            f" stdout: {output.stdout} \n stderr :{output.stderr}")
+    return output
+
+
+def set_param_cutoff(cache_id: int, core_id: int = None, threshold: Size = None,
+                     policy: SeqCutOffPolicy = None):
+    if core_id is None:
+        command = set_param_cutoff_cmd(
+            cache_id=str(cache_id), threshold=threshold.get_value(Unit.KibiByte), policy=policy.name)
```
[the remark this answers is not in the document]
Comment (KamilLepek) [reply api/cas/casadm.py]: Done

**Pull request merged**